docs/tutorials/go-built-in.md

```diff
@@ -103,579 +101,527 @@ func NewKVStoreApplication() *KVStoreApplication {
+
+`go get github.com/dgraph-io/badger`
+
+Next, let's update the application and its constructor to receive a handle to the 
+database. 
 
-When a new transaction is added to the Tendermint Core, it will ask the
-application to check it (validate the format, signatures, etc.).
+Update the application struct as follows: 
 
 ```go
-import "bytes"
-
-func (app *KVStoreApplication) isValid(tx []byte) (code uint32) {
- // check format
- parts := bytes.Split(tx, []byte("="))
- if len(parts) != 2 {
-  return 1
- }
-
- key, value := parts[0], parts[1]
-
- // check if the same key=value already exists
- err := app.db.View(func(txn *badger.Txn) error {
-  item, err := txn.Get(key)
-  if err != nil && err != badger.ErrKeyNotFound {
-   return err
-  }
-  if err == nil {
-   return item.Value(func(val []byte) error {
-    if bytes.Equal(val, value) {
-     code = 2
-    }
-    return nil
-   })
-  }
-  return nil
- })
- if err != nil {
-  panic(err)
- }
-
- return code
+type KVStoreApplication struct {
+	db           *badger.DB
+	currentBatch *badger.Txn
 }
+```
 
-func (app *KVStoreApplication) CheckTx(req abcitypes.RequestCheckTx) abcitypes.ResponseCheckTx {
- code := app.isValid(req.Tx)
- return abcitypes.ResponseCheckTx{Code: code, GasWanted: 1}
+And change the constructor to set the appropriate field when creating the application:
+
+```go
+func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
+	return &KVStoreApplication{
+		db: db,
+	}
 }
 ```
 
-Don't worry if this does not compile yet.
-
-If the transaction does not have a form of `{bytes}={bytes}`, we return `1`
-code. When the same key=value already exist (same key and value), we return `2`
-code. For others, we return a zero code indicating that they are valid.
+Don't worry about `currentBatch` for now, we'll get to that later.
 
-Note that anything with non-zero code will be considered invalid (`-1`, `100`,
-etc.) by Tendermint Core.
+### 1.3.1 CheckTx
 
-Valid transactions will eventually be committed given they are not too big and
-have enough gas. To learn more about gas, check out ["the
-specification"](https://docs.tendermint.com/master/spec/abci/apps.html#gas).
+When Tendermint Core receives a new transaction, Tendermint asks the application
+if the transaction is acceptable. In our new application, let's implement some
+basic validation for the transactions it will receive.
 
-For the underlying key-value store we'll use
-[badger](https://github.com/dgraph-io/badger), which is an embeddable,
-persistent and fast key-value (KV) database.
+Add the following helper method to `app.go`:
 
 ```go
-import "github.com/dgraph-io/badger"
+func (app *KVStoreApplication) validateTx(tx []byte) uint32 {
+	parts := bytes.Split(tx, []byte("="))
+
+	// first, check that the transaction is not malformed.
+	if len(parts) != 2 {
+		return 1
+	}
+
+	key, value := parts[0], parts[1]
+
+	var txAlreadyExists bool
+
+	// ok, we have a well-formed transaction, next let's check that we haven't
```

Comment:
Hmm. I'm not sure we really want to do this. Shouldn't it be possible to update a key that was previously written? (Maybe not within the same block, but that's not really what we're checking here—this scans the already-stored data, not the pending dbtx).